src/backend/distributed/connection/connection_management.c

```diff
@@ -527,19 +535,207 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
+			RemoteSocketClosed(connection))
+		{
+			/*
+			 * If a connection is not usable and we do not need to throw an error,
+			 * close at the end of the transaction.
+			 */
+			connection->forceCloseAtTransactionEnd = true;
+		}
+		else
+		{
+			/* connection is healthy */
+			healthyConnections = lappend(healthyConnections, connection);
+		}
+	}
+
+	return healthyConnections;
+}
+
+
+/*
+ * RemoteSocketClosed returns true if the remote socket is closed.
+ *
+ * Note that we rely on WL_SOCKET_CLOSED, which is only available
+ * for PG 15+ and for kernels that support the feature.
+ */
+static bool
+RemoteSocketClosed(MultiConnection *connection)
+{
+	bool socketClosed = false;
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+	if (!WaitEventSetCanReportClosed())
+	{
+		/* the kernel build does not support WL_SOCKET_CLOSED */
+		return false;
+	}
+
+	WaitEvent events[3];
+	WaitEventSet *waitEventSet =
+		CreateWaitEventSet(CurrentMemoryContext, 3);
+	int sock = PQsocket(connection->pgConn);
+	int waitEventSetIndex =
+		CitusAddWaitEventSetToSet(waitEventSet, WL_SOCKET_CLOSED, sock,
+								  NULL, (void *) connection);
+
+	if (waitEventSetIndex == WAIT_EVENT_SET_INDEX_FAILED)
+	{
+		/*
+		 * Inform failed to add to wait event set with a debug message as this
+		 * is too detailed information for users. We let the code flow just in
+		 * case it can successfully finish the execution.
+		 */
+		ereport(DEBUG1, (errcode(ERRCODE_CONNECTION_FAILURE),
+						 errmsg("Adding wait event for node %s:%d failed. "
+								"The socket was: %d",
+								connection->hostname,
+								connection->port, sock)));
+
+		FreeWaitEventSet(waitEventSet);
+
+		return false;
+	}
+
+	CitusAddWaitEventSetToSet(waitEventSet, WL_POSTMASTER_DEATH, PGINVALID_SOCKET, NULL,
+							  NULL);
+	CitusAddWaitEventSetToSet(waitEventSet, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch,
+							  NULL);
+
+	bool retry = false;
+	do {
+		long timeout = 0; /* do not wait at all */
+		int eventCount =
+			WaitEventSetWait(waitEventSet, timeout, events,
+							 3, WAIT_EVENT_CLIENT_READ);
+
+		retry = ProcessWaitEventsForSocketClose(events, eventCount, &socketClosed);
+		if (socketClosed)
+		{
+			/* socket is closed, no further information needed */
+			break;
+		}
+	} while (retry);
+
+
+	FreeWaitEventSet(waitEventSet);
+#endif
+
+	return socketClosed;
+}
+
+
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+/*
+ * ProcessWaitEventsForSocketClose goes over the events and mainly looks for
+ * WL_SOCKET_CLOSED event. When found, socketClosed is set to true.
+ *
+ * The function also returns true if we are required to re-check the events. We
+ * might need to re-check if latch is set. A latch event might be preventing
+ * other events from being reported. Poll and check again.
+ */
+static bool
+ProcessWaitEventsForSocketClose(WaitEvent *events, int eventCount, bool *socketClosed)
+{
+	*socketClosed = false;
+
+	int eventIndex = 0;
+	for (; eventIndex < eventCount; eventIndex++)
```

Comment:
Now that we are in our own function maybe also do the more general thing for allocate/initialize the index.

```suggestion
	for (int eventIndex = 0; eventIndex < eventCount; eventIndex++)
```